yjit/src/compilation_log.rs

```diff
@@ -0,0 +1,219 @@
+use crate::core::BlockId;
+use crate::cruby::*;
+use crate::options::*;
+use crate::yjit::yjit_enabled_p;
+
+use std::fmt::{Display, Formatter};
+use std::os::raw::c_long;
+use crate::utils::iseq_get_location;
+
+type Timestamp = f64;
+
+#[derive(Clone, Debug)]
+pub struct CompilationLogEntry {
+    /// The time when the block was compiled.
+    pub timestamp: Timestamp,
+
+    /// The compilation log message.
+    pub message: String,
+}
+
+impl Display for CompilationLogEntry {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:15.6}: {}", self.timestamp, self.message)
+    }
+}
+
+pub type CompilationLog = CircularBuffer<CompilationLogEntry, 1024>;
+static mut COMPILATION_LOG : Option<CompilationLog> = None;
+
+impl CompilationLog {
+    pub fn init() {
+        unsafe {
+            COMPILATION_LOG = Some(CompilationLog::new());
+        }
+    }
+
+    pub fn get_instance() -> &'static mut CompilationLog {
+        unsafe {
+            COMPILATION_LOG.as_mut().unwrap()
+        }
+    }
+
+    pub fn has_instance() -> bool {
+        unsafe {
+            COMPILATION_LOG.as_mut().is_some()
+        }
+    }
+
+    pub fn add_block_with_chain_depth(block_id: BlockId, chain_depth: u8) {
+        if !Self::has_instance() {
+            return;
+        }
+
+        let timestamp = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap().as_secs_f64();
+
+        let location = iseq_get_location(block_id.iseq, block_id.idx);
+        let message = if chain_depth > 0 {
+            format!("{} (chain_depth: {})", location, chain_depth)
+        } else {
+            format!("{}", location)
+        };
+
+        let entry = CompilationLogEntry {
+            timestamp,
+            message
+        };
+
+        match get_option!(compilation_log) {
+            CompilationLogOutput::Stderr => {
+                eprintln!("{}", entry);
+            }
+
+            CompilationLogOutput::File(fd) => {
+                use std::os::unix::io::{FromRawFd, IntoRawFd};
+                use std::io::Write;
+
+                // Write with the fd opened during boot
+                let mut file = unsafe { std::fs::File::from_raw_fd(fd) };
+                writeln!(file, "{}", entry).unwrap();
+                file.flush().unwrap();
+                file.into_raw_fd(); // keep the fd open
+            }
+
+            _ => () // Don't write/print anything
+        }
+
+        Self::get_instance().push(entry);
+    }
+
+    pub fn clear() {
+        unsafe {
+            COMPILATION_LOG.as_mut().unwrap().reset()
+        }
+    }
+}
+
+pub struct CircularBuffer<T, const N: usize> {
```

Comment:
Can you explain why you chose to implement this instead of using [std::collections::VecDeque](https://doc.rust-lang.org/std/collections/struct.VecDeque.html)?

Reply:
Up until I made adjustments from review feedback the circular buffer was a fixed size. I was under the impression that's what we wanted. We can change to a standard collection easily enough.

Reply:
A requirement for this project was to bound the size of the collection, with the collection always containing the N most recent log entries (specifically, a circular buffer). Looking at the `VecDeque` docs, I don't see a way to use a `VecDeque` in such a way. I could layer the semantics on top of that, but I think we're largely back to the current implementation.

I'd prefer not rolling my own, but we can't add dependencies either. If the concern is the `clone()`, that really shouldn't have been there. I was just trying to get the iterator working when I switched the payload out to `String`. I can remove that easily enough. I suppose the easiest thing is to just use `pop()`. The buffer is already being cleared once fetched by Ruby. The most straightforward would be:

```rust
 while log.len() > 0 {
        let entry = log.pop().unwrap();

        unsafe {
            let entry_array = rb_ary_new_capa(2);
            rb_ary_push(entry_array, rb_float_new(entry.timestamp));
            rb_ary_push(entry_array, entry.message.into());
            rb_ary_push(array, entry_array);
        }
    }
```

I could update the iterator's `next()` to call `pop()`. That'd let us use a `for` loop, which may look a bit more idiomatic. But, it may be unconventional for the iterator to mutate the backing store. In this case, the store isn't being physically resized so we don't have to deal with moving memory locations.

Do you have a preferred path?

Reply:
I've gone with the `while` loop for now, but it's a separate commit that can be dropped easily. If that's the way we go I can squash it back into the original and remove the iterator entirely.

Reply:
I'm more okay with hand-rolling this with the iterator gone. 👍🏼